Clarify Git Flow usage and installation.

alastairs opened the pull request:
Add further instructions as per [this comment](https://github.com/GiveCampUK/GiveCRM/pull/94#issuecomment-5006645) to clarify that usage of Git Flow is not required for contributing to GiveCRM, and to offer a bit more help with installation of Git Flow.

If there are any more comments, or stuff that I've missed, add them to this thread and leave the Pull Request open; I'll update it with new commits as I address the comments.

Ta

Alastair

alastairs committed:
Clarify Git Flow usage and installation.

Add further instructions as per [this
comment](https://github.com/GiveCampUK/GiveCRM/pull/94#issuecomment-5006645)
to clarify that usage of Git Flow is not required for contributing to
GiveCRM, and to offer a bit more help with installation of Git Flow.

nbarnwell commented on ReadMe.md:
```diff
@@ -27,11 +27,27 @@ We'll do a bit of a code review before accepting your patch.
+
+Pick a feature or bug to work on and create a new branch for that work by typing `git flow feature start <featurename>`.  This will create you a new branch for your work, and you can use git as usual from this point.  
+
+Once your feature is finished, type `git flow feature publish <featurename>`.  This will push the feature up to your `origin` repository on GitHub and you will then be able to submit a pull request to have it merged into `develop`.  **Note: do not use `git flow feature finish <featurename>`!**  This will automatically merge your feature branch back into `develop` and delete the feature branch, ma ...
```
Could you just clarify whether this is "origin/master" or "origin/develop", please? In fact, it would be ace if throughout this document we used a convention that tells the reader which branch in which repo and whether it's theirs or GiveCRM's. e.g. (username|givecrm)/reponame/branchname?

alastairs replied:
Ok, I misread this on first blush, and thought you were referring to the `git flow feature publish` command, but now I think you're referring to the warning about `git flow feature finish`.  I'll put together a more convention-based description and update the PR.

alastairs replied:
That said, git commands only operate on the local repository, with the exception of push.  So `git merge` could not modify GiveCampUK/GiveCRM unless you explicitly pushed changes made to the local branch to that remote repository.

nbarnwell replied:
I was referring to where you say "This will push the feature up to your 'origin' repository...". Do you mean that `git flow [feature] publish` will **push** changes to `nbarnwell/master` or `nbarnwell/develop`? Totally cool if that's what happens, but is it repeatable? I.e. can I run it more than once before opening the pull request? I mean, what if I want to push to Github several times before the feature is technically worth "publishing"? Sorry for being such a n00b. :S

alastairs replied:
No worries, it's kinda hard to explain with three different repositories in the picture: local "working copy" clone (the one on your computer), remote "GitHub"/"fork" clone (github.com/username/GiveCRM), and upstream "central" repository (github.com/GiveCampUK/GiveCRM).  

The answer is neither: it will push changes to `nbarnwell/feature/<featurename>`. Unfortunately, you can't run `git flow feature publish` more than once; after you've done it once, you have to use the usual `git push origin` to update the published branch.

nbarnwell replied:
Okay that makes sense, now. So I push feature branches to `origin/feature/<featurename>` (initially using `git flow feature publish <featurename>` but subsequently using vanilla `git push origin feature/<featurename>`), and submit a pull request against that branch.

Of someone else does that, do we merge their feature branch into our main `develop` branch, then (i.e. we don't want the main repo history full of people's deleted feature/hotfix branches)? If so, how? I'm assuming that rules out use of Github's automated pull request merging?

alastairs replied:
Yes, that's what we do, but no it doesn't rule out GitHub's auto-merging feature.  Pull Requests are raised against `develop`, by virtue of that branch now being the default for the repository; this is what happened with all my pull requests in the last 1-2 weeks.  

A note on the branching strategy, which might make things a bit clearer (or just confuse the issue further!): the `develop` branch moves ahead of `master` with each merged pull request, and will consist only of these pull request merge commits.  When we're ready to ship a production release, `develop` will be merged into `master`, which will be a nice fast-forward merge. We can then optionally tag that release as well, and `develop` will start to move ahead of `master` again with newly-submitted pull requests.

alastairs committed:
Reword section on git flow feature publish.

Introduce the "feature branch" terminology, and format its usage
consistently across the section.

Describe git flow feature publish as "copying" the feature branch to
your origin repository, rather than pushing it.  Hopefully this makes it
a bit easier to understand for people new to git (and git flow), whilst
still making it obvious to experienced git users what is going on under
the hood.

nbarnwell commented on ReadMe.md:
```diff
@@ -27,11 +27,27 @@ We'll do a bit of a code review before accepting your patch.
-Once your feature is finished, type `git flow feature publish <featurename>`.  This will push the feature up to your `origin` repository on GitHub and you will then be able to submit a pull request to have it merged into `develop`.  
+#### Configuring Git Flow
+
+There is a set of [helper scripts](https://github.com/nvie/gitflow) that will work on both Unix-based operating systems and Windows.  Follow the appropriate [installation instructions](https://github.com/nvie/gitflow/wiki/Installation) for your operating system, and configure your working copy repository for use with Git Flow by typing `git flow init`.  
```
`git flow init` asks me a bunch of questions - should I just choose the default values it suggests?

alastairs replied:
I'm assuming you edited this comment, as it's different from the email notification I received...

Yes: choose `master` for production releases, `develop` for continuing development, and use the defaults for the various prefixes for features, hotfixes, etc.

alastairs committed:
Clarify usage of `git flow init`.

Accept all default answers to the various questions asked.  Also include
instructions on updating the published feature branch.

alastairs committed:
Correctly format a reference to the develop branch.
